### ISSUE TITLE
'vendor' is not recognized as an internal or external command, operable program or batch file.

### ISSUE THREAD

**Issue by lazychaser** [lorisleiva/laravel-deployer#22]
Running on windows.

**Comment by lorisleiva** [comment]
Mais encore... Could you provide more context?

**Comment by lazychaser** [comment]
This happens when I'm trying to run any artisan command, It is trying to start a new process `vendor/bin/dep` and fails. Replacing `/` with `DIRECTORY_SEPARATOR` helps, but now it says `Command "deploy" is not defined.`. I tried to run deployer manually and it says the same thing.

**Comment by lorisleiva** [comment]
What version of Deployer do you end up using? Did you have it installed before using Laravel Deployer? Do you have access to other binaries in your vendor folder? Can you also show me the code of your “/“ replacement?

**Comment by lazychaser** [comment]
I never used Deployer before, I simply required your package. This was on L5.5, now I'm upgrading to L5.6.

The replacement was following:

```
$ds = DIRECTORY_SEPARATOR;
$this->process("vendor{$ds}bin{$ds}dep ...");
```

This allowed me to start a proccess, but is semeed like it was not able to discover `laravel-deployer.php` file that was in compiled `deploy.php`.

**Comment by lorisleiva** [comment]
Okay then I think the problem is due to other default paths that are not using the DIRECTORY_SEPARATOR variable. I will fix that today and come back to you ASAP. Thank you for reporting this.

**Comment by lorisleiva** [comment]
I made Laravel Deployer use `DIRECTORY_SEPARATOR` instead of `/` on strategic places of the source code. Could you please upgrade Laravel Deployer to `dev-master` and tell me if this fixes your issue? If it doesn't please provide the output of `php artisan deploy -vvv`.

**Comment by lazychaser** [comment]
```
$ php artisan deploy -vvv

In Application.php line 632:

  [Symfony\Component\Console\Exception\CommandNotFoundException]
  Command "deploy" is not defined.


Exception trace:
 Symfony\Component\Console\Application->find() at C:\www\asmo\chains\vendor\symfony\console\Application.php:226
 Symfony\Component\Console\Application->doRun() at C:\www\asmo\chains\vendor\symfony\console\Application.php:145
 Symfony\Component\Console\Application->run() at C:\www\asmo\chains\vendor\deployer\deployer\src\Deployer.php:331
 Deployer\Deployer::run() at C:\www\asmo\chains\vendor\deployer\deployer\bin\dep:120


In Process.php line 223:

  [Symfony\Component\Process\Exception\ProcessFailedException]
  The command "vendor\bin\dep --file='C:\www\asmo\chains\vendor\lorisleiva\laravel-deployer\src\LaravelDeployer\..\..\.build\deploy.php' deploy --verbose -vvv" failed.

  Exit Code: 1(General error)

  Working directory: C:\www\asmo\chains

  Output:
  ================


  Error Output:
  ================

  In Application.php line 632:

    [Symfony\Component\Console\Exception\CommandNotFoundException]
    Command "deploy" is not defined.


  Exception trace:
   Symfony\Component\Console\Application->find() at C:\www\asmo\chains\vendor\symfony\console\Application.php:226
   Symfony\Component\Console\Application->doRun() at C:\www\asmo\chains\vendor\symfony\console\Application.php:145
   Symfony\Component\Console\Application->run() at C:\www\asmo\chains\vendor\deployer\deployer\src\Deployer.php:331
   Deployer\Deployer::run() at C:\www\asmo\chains\vendor\deployer\deployer\bin\dep:120


Exception trace:
 Symfony\Component\Process\Process->mustRun() at C:\www\asmo\chains\vendor\lorisleiva\laravel-deployer\src\LaravelDeployer\Commands\BaseCommand.php:110
 Lorisleiva\LaravelDeployer\Commands\BaseCommand->process() at C:\www\asmo\chains\vendor\lorisleiva\laravel-deployer\src\LaravelDeployer\Commands\BaseCommand.php:58
 Lorisleiva\LaravelDeployer\Commands\BaseCommand->dep() at C:\www\asmo\chains\vendor\lorisleiva\laravel-deployer\src\LaravelDeployer\Commands\Deploy.php:12
 Lorisleiva\LaravelDeployer\Commands\Deploy->handle() at n/a:n/a
 call_user_func_array() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Container\BoundMethod.php:29
 Illuminate\Container\BoundMethod::Illuminate\Container\{closure}() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Container\BoundMethod.php:87
 Illuminate\Container\BoundMethod::callBoundMethod() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Container\BoundMethod.php:31
 Illuminate\Container\BoundMethod::call() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Container\Container.php:564
 Illuminate\Container\Container->call() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Console\Command.php:184
 Illuminate\Console\Command->execute() at C:\www\asmo\chains\vendor\symfony\console\Command\Command.php:251
 Symfony\Component\Console\Command\Command->run() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Console\Command.php:171
 Illuminate\Console\Command->run() at C:\www\asmo\chains\vendor\symfony\console\Application.php:886
 Symfony\Component\Console\Application->doRunCommand() at C:\www\asmo\chains\vendor\symfony\console\Application.php:262
 Symfony\Component\Console\Application->doRun() at C:\www\asmo\chains\vendor\symfony\console\Application.php:145
 Symfony\Component\Console\Application->run() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Console\Application.php:89
 Illuminate\Console\Application->run() at C:\www\asmo\chains\vendor\laravel\framework\src\Illuminate\Foundation\Console\Kernel.php:122
 Illuminate\Foundation\Console\Kernel->handle() at C:\www\asmo\chains\artisan:37
```

**Comment by lorisleiva** [comment]
Could you please tell me which version of Deployer you are using (in your composer.lock)?

**Comment by lazychaser** [comment]
v6.2.0

**Comment by lorisleiva** [comment]
Can you show me your `config/deploy.php`?
Can you show me the content of the file located in `vendor/lorisleiva/laravel-deployer/.build/deploy.php`?

**Comment by lazychaser** [comment]
Pretty basic setup after `deploy:init`.

```php
<?php

namespace Deployer;

require 'recipe/laravel-deployer.php';

/*
 * Includes
 */

/*
 * Options
 */

set('strategy', 'basic');
set('application', 'Laravel');
set('repository', 'root@sms47.com:/var/git/chains.git');
set('shared_dirs', [
    'storage',
    'public/static',
    'public/.well-known',
]);
set('branch', 'test/deployer');

/*
 * Hosts and localhost
 */

host('sms47.com')
    ->set('deploy_path', '/var/www/m.bot-marketing.com')
    ->user('root');

/*
 * Strategies
 */

/*
 * Hooks
 */

after('hook:build', 'npm:install');
after('hook:build', 'npm:production');
after('hook:ready', 'artisan:storage:link');
after('hook:ready', 'artisan:view:clear');
after('hook:ready', 'artisan:cache:clear');
after('hook:ready', 'artisan:config:cache');
after('hook:ready', 'artisan:route:cache');
after('hook:ready', 'artisan:migrate');
```

```php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Default deployment strategy
    |--------------------------------------------------------------------------
    |
    | This option defines which deployment strategy to use by default on all
    | of your hosts. Laravel Deployer provides some strategies out-of-box
    | for you to choose from explained in detail in the documentation.
    |
    | Supported: 'basic', 'firstdeploy', 'local', 'pull'.
    |
    */

    'default' => 'basic',

    /*
    |--------------------------------------------------------------------------
    | Custom deployment strategies
    |--------------------------------------------------------------------------
    |
    | Here, you can easily set up new custom strategies as a list of tasks.
    | Any key of this array are supported in the `default` option above.
    | Any key matching Laravel Deployer's strategies overrides them.
    |
    */

    'strategies' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Hooks
    |--------------------------------------------------------------------------
    |
    | Hooks let you customize your deployments conveniently by pushing tasks
    | into strategic places of your deployment flow. Each of the official
    | strategies invoke hooks in different ways to implement their logic.
    |
    */

    'hooks' => [
        // Right before we start deploying.
        'start' => [
            //
        ],

        // Code and composer vendors are ready but nothing is built.
        'build' => [
            'npm:install',
            'npm:production',
        ],

        // Deployment is done but not live yet (before symlink)
        'ready' => [
            'artisan:storage:link',
            'artisan:view:clear',
            'artisan:cache:clear',
            'artisan:config:cache',
            'artisan:route:cache',
            'artisan:migrate',
        ],

        // Deployment is done and live
        'done' => [
            //
        ],

        // Deployment succeeded.
        'success' => [
            //
        ],

        // Deployment failed.
        'fail' => [
            //
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Deployment options
    |--------------------------------------------------------------------------
    |
    | Options follow a simple key/value structure and are used within tasks
    | to make them more configurable and reusable. You can use options to
    | configure existing tasks or to use whithin your own custom tasks.
    |
    */

    'options' => [
        'application' => env('APP_NAME', 'Laravel'),
        'repository' => 'root@sms47.com:/var/git/chains.git',
        'shared_dirs' => [ 'storage', 'public/static', 'public/.well-known' ],
        'branch' => 'test/deployer',
    ],

    /*
    |--------------------------------------------------------------------------
    | Hosts
    |--------------------------------------------------------------------------
    |
    | Here, you can define any domain or subdomain you want to deploy to.
    | You can provide them with roles and stages to filter them during
    | deployment. Read more about how to configure them in the docs.
    |
    */

    'hosts' => [
        'sms47.com' => [
            'deploy_path' => '/var/www/m.bot-marketing.com',
            'user' => 'root',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Localhost
    |--------------------------------------------------------------------------
    |
    | This localhost option give you the ability to deploy directly on your
    | local machine, without needing any SSH connection. You can use the
    | same configurations used by hosts to configure your localhost.
    |
    */

    'localhost' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Include additional Deployer recipes
    |--------------------------------------------------------------------------
    |
    | Here, you can add any third party recipes to provide additional tasks,
    | options and strategies. Therefore, it also allows you to create and
    | include your own recipes to define more complex deployment flows.
    |
    */

    'include' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Use a custom Deployer file
    |--------------------------------------------------------------------------
    |
    | If you know what you are doing and want to take complete control over
    | Deployer's file, you can provide its path here. Note that, without
    | this configuration file, the root's deployer file will be used.
    |
    */

    'custom_deployer_file' => false,

];
```

**Comment by lorisleiva** [comment]
What happens when you run `vendor/bin/dep --file='vendor/lorisleiva/laravel-deployer/.build/deploy.php' list`?

**Comment by lazychaser** [comment]
It works fine! Displays all tasks.

**Comment by lorisleiva** [comment]
It looks like the problem comes from the double dots `..\..` on the absolute path `C:\www\asmo\chains\vendor\lorisleiva\laravel-deployer\src\LaravelDeployer\..\..\.build\deploy.php` on Windows.

I am going to try to use `dirname` to see if it makes a difference.

**Comment by lazychaser** [comment]
Maybe `realpath`?

**Comment by lorisleiva** [comment]
Sorry yes I meant `realpath`. Could you please update to the latest commit and try again?

**Comment by lazychaser** [comment]
I managed to figure it out. The problem is in deployer. This is in `vendor/deployer/deployer/bin/dep` file:

```php
$deployFile = ($userSpecifiedFile[0] === '/' ? '' : getcwd() . '/') . $userSpecifiedFile;
```

It appends current working directory if path doesn't start with `/`. In Windows, path never starts with it. So we need to pass relative path.

I changed `store` method of `DeployFile`:

```php
public function store()
    {
        $ds = DIRECTORY_SEPARATOR;
        $path = "vendor{$ds}lorisleiva{$ds}laravel-deployer{$ds}.build{$ds}deploy.php";
        $dir = dirname($path);

        if (! is_dir($dir)) {
            mkdir($dir, 0777, true);
        }

        $this->filesystem->put($path, (string) $this);

        return $path;
    }
```

And I also had to remove `'` in `BaseCommand`:

```
$this->process("$depBinary --file=$deployFile $command $parameters");
```

This allowed me to list all the tasks.

**Comment by lorisleiva** [comment]
Okay thanks for that. Does it not work with the latest commit that uses `realpath` then?

**Comment by lorisleiva** [comment]
Okay I've now committed your changes. Could you upgrade to the latest commit and confirm that everything works by closing this issue?

**Comment by lazychaser** [comment]
Everything is ok. Waiting for release.

**Comment by lorisleiva** [comment]
@lazychaser Just to let you know that these changes are now available in [release v0.2.4](https://github.com/lorisleiva/laravel-deployer/releases/tag/v0.2.4)
